src/medCore/gui/viewContainers/medViewContainer.cpp

```diff
@@ -741,13 +757,37 @@ void medViewContainer::openFromSystem()
     if (path.isEmpty())
         return;
 
-    connect(medDataManager::instance(), SIGNAL(dataImported(medDataIndex,QUuid)), this, SLOT(dataReady(medDataIndex,QUuid)));
-    d->expectedUuid = medDataManager::instance()->importPath(path, true, false);
+    open(path);
+}
 
+void medViewContainer::open(const QString & path)
+{
+    QEventLoop loop;
+    QUuid uuid = medDataManager::instance()->importPath(path, false);
+    d->expectedUuids.append(uuid);
+    connect(medDataManager::instance(), SIGNAL(dataImported(medDataIndex,QUuid)),
+            this, SLOT(open_waitForImportedSignal(medDataIndex,QUuid)));
+    while( d->expectedUuids.contains(uuid)) {
+        loop.processEvents(QEventLoop::ExcludeUserInputEvents);
```

Comment:
There are some multithreading issues here, and in particular the access to ```expectedUuids``` throughout the code is not guarded by a mutex.
Here is how it's done in ```medPipelineWorkspace```:

The private class:
```
class medPipelineWorkspacePrivate
{
public:
    (...)
    QMutex databaseMutex;
    QHash<QUuid, dtkSmartPointer<medAbstractData> > currentImports;
    (...)

    medPipelineWorkspacePrivate() : databaseMutex(QMutex::Recursive) {}
};
```

In the constructor:
```
    connect(medDataManager::instance(), SIGNAL(dataImported(medDataIndex, QUuid)),
            this, SLOT(dataImported(medDataIndex, QUuid)));
```

The function to call the import:
```
void medPipelineWorkspace::secureImport(dtkSmartPointer<medAbstractData> data)
{
    d->databaseMutex.lock();
    QUuid importId = medDataManager::instance()->importData(data, true);
    d->currentImports[importId] = data;
    d->databaseMutex.unlock();

    QEventLoop loop;
    connect(this, SIGNAL(importFinished()), &loop, SLOT(quit()));
    loop.exec();
}
```

The slot that is called on termination of the import:
```
void medPipelineWorkspace::dataImported(const medDataIndex& index, QUuid importId)
{
    QMutexLocker(&d->databaseMutex);

    if (d->currentImports.contains(importId))
    {
        dtkSmartPointer<medAbstractData> data = d->currentImports.take(importId);
        data->setDataIndex(index);
        emit importFinished();
    }
}
```

I called it secure import because at the time its goal was to prevent the situation where you deleted data just after the call to import (such as a smartPointer going out of scope), which caused a crash. That's why it stores not only a uuid but also a pointer to the data. In your case you don't need that aspect so you can replace the hash with a list of uuids like you already have.